lib/eventFactory.js

```diff
@@ -2,9 +2,94 @@
 
 const BaseFactory = require('./baseFactory');
 const Event = require('./event');
+const jobConfigSchema = require('screwdriver-data-schema').config.job;
+const joi = require('joi');
 
 let instance;
 
+/**
+ * Create builds associated with this event
+ * For example, if startFrom ~commit, then create builds with jobs that have requires: ~commit
+ * @method createBuilds
+ * @param  {Pipeline} pipeline                   Pipeline to create builds
+ * @param  {Object}   config
+ * @param  {String}   config.sha                 SHA this project was built on
+ * @param  {String}   config.username            Username of the user that creates this event
+ * @param  {String}   config.scmContext          The scm context to which user belongs
+ * @param  {String}   [config.prRef]             Ref if it's a PR event
+ * @param  {String}   [config.prNum]             PR Number if it's a PR event
+ * @param  {String}   [config.startFrom]         Where the event starts from (jobname or ~commit, ~pr, etc)
+ *                                               Making optional for backward compatibility
+ * @param  {Number}   eventId                    Event's id
+ */
+function createBuilds(pipeline, config, eventId) {
+    const startFrom = config.startFrom;
+
+    // If no startFrom, then do nothing. Remove once we switch to new workflow design
+    if (!startFrom) {
+        return null;
+    }
+
+    /* eslint-disable global-require */
+    const BuildFactory = require('./buildFactory');
+    const buildFactory = BuildFactory.getInstance();
+    const JobFactory = require('./jobFactory');
+    const jobFactory = JobFactory.getInstance();
+    /* eslint-enable global-require */
+
+    const createJob = (job, jobName) => jobFactory.create({
+        pipelineId: job.pipelineId,
+        name: jobName,
+        permutations: job.permutations
+    });
+
+    return pipeline.jobs.then((jobs) => {
+        const jobNamesArray = jobs.map(j => j.name);
+        const jobsToCreate = [];
+        let jobsToStart = [];
+
+        jobs.forEach((job) => {
+            const enabled = job.state === 'ENABLED';
+            const requires = job.permutations.requires;
+            const isJobName = joi.validate(startFrom, jobConfigSchema.jobname);
```

Comment:
I would rather have `const isJobName = !!joi.validate(startFrom, jobConfigSchema.jobname).error`

and then `isJobName && startFrom === job.name` This way if conditions in line 59 & 62 are more readable